src/System.IO.Pipes/tests/NamedPipesSimpleTest.cs

```diff
@@ -224,4 +339,134 @@ public static void ClientPInvokeChecks()
             }
         }
     }
+
+    [Fact]
+    [PlatformSpecific(PlatformID.Windows)]
+    public static void ClientServerMessages()
+    {
+        byte[] msg1 = new byte[] { 5, 7, 9, 10 };
+        byte[] msg2 = new byte[] { 2, 4 };
+        byte[] received1 = new byte[] { 0, 0, 0, 0 };
+        byte[] received2 = new byte[] { 0, 0 };
+        byte[] received3 = new byte[] { 0, 0, 0, 0}; ;
+
+        using (NamedPipeServerStream server = new NamedPipeServerStream("foomsg", PipeDirection.InOut, 1, PipeTransmissionMode.Message))
+        {
+            using (NamedPipeClientStream client = new NamedPipeClientStream(".", "foomsg", PipeDirection.InOut, PipeOptions.None, System.Security.Principal.TokenImpersonationLevel.Identification))
+            {
+                Task clientTask = Task.Run(() =>
+                {
+                    client.Connect();
+                    
+                    client.Write(msg1, 0, msg1.Length);
+                    client.Write(msg2, 0, msg2.Length);
+                    client.Write(msg1, 0, msg1.Length);
+
+                    int serverCount = client.NumberOfServerInstances;
+                    Assert.Equal(1, serverCount);
+                });
+
+                server.WaitForConnection();
+                int len1 = server.Read(received1, 0, msg1.Length);
+                Assert.True(server.IsMessageComplete);
+                Assert.Equal(msg1.Length, len1);
+                Assert.Equal(msg1, received1);
+
+                int len2 = server.Read(received2, 0, msg2.Length);
+                Assert.True(server.IsMessageComplete);
+                Assert.Equal(msg2.Length, len2);
+                Assert.Equal(msg2, received2);
+
+                int len3 = server.Read(received3, 0, msg1.Length - 1);  // read one less than message
+                Assert.False(server.IsMessageComplete);
+                Assert.Equal(msg1.Length - 1, len3);
+
+                int len4 = server.Read(received3, len3, received3.Length - len3);
+                Assert.True(server.IsMessageComplete);
+                Assert.Equal(msg1.Length, len3 + len4);
+                Assert.Equal(msg1, received3);
+
+                string userName = server.GetImpersonationUserName();    // not sure what to test here that will work in all cases
+            }
+        }
+    }
+
+    [Fact(Skip = "test hangs right now")]
```

Comment:
If you believe this is a product bug, you can open an issue for it on GitHub, and then use:
[ActiveIssue(TheIssueNumberHere)]